### PR TITLE
feat: implement online data fetching

### PR DESCRIPTION
**Pull request by Thepowa753**
for those who don't have the hardware to receive the packets themselves

### CODE REVIEW

File: app/settings/__init__.py [hunk 1]
```diff
@@ -13,6 +13,7 @@ class Settings(BaseModel):
     listener_location: ListenerLocation
     notification_thresholds: NotificationThresholds
     udp_broadcast: UDPBroadcast
+    fetch_data: bool = Field(default=True)
```
Comment (ch3p4ll3) [inline app/settings/__init__.py:16]: the name of this variable is not very clear, it should be changed to something else. like “fetch_from_online”

File: app/radiosonde_auto_rx_listener.py [hunk 2]
```diff
@@ -21,10 +21,97 @@ def __init__(self):
...
+        if self._settings.fetch_data:
+            logger.info("Fetching data from online sites.")
+            await self._listen_web()
+        else:
+            logger.info("Fetching data from local udp.")
+            await self._listen_udp()
+
+    async def _listen_web(self):
+        """Uses async requests to fetch data from online sources."""
+        url = f"https://s1.radiosondy.info/export/export_map.php?live_map=1&_={int(datetime.now().timestamp() * 1000)}"
```
Comment (ch3p4ll3) [inline app/radiosonde_auto_rx_listener.py:35]: I think the urls must be updated ad each iteration with the new timestamp, or not?

File: app/radiosonde_auto_rx_listener.py [hunk 3]
```diff
@@ -21,10 +21,97 @@ def __init__(self):
...
+
+        headers = {
```
Comment (ch3p4ll3) [inline app/radiosonde_auto_rx_listener.py:37]: u sure we need all this?

File: app/radiosonde_auto_rx_listener.py [hunk 4]
```diff
@@ -21,10 +21,97 @@ def __init__(self):
...
+                        for feature in data["features"]:
+                            if (
+                                "geometry" in feature
+                                and feature["geometry"]["type"] == "Point"
+                            ):
+                                coords = (
+                                    feature["geometry"]["coordinates"][1],
+                                    feature["geometry"]["coordinates"][0],
+                                )
+                                if Utils.is_within_range(
```
Comment (ch3p4ll3) [inline app/radiosonde_auto_rx_listener.py:77]: there is no logic to stop the script from spamming notifications... for example: if a radiosonde meets all criteria each time you grab the new data, the script will send a notification. see the use os self._sondes :)

File: app/radiosonde_auto_rx_listener.py [hunk 5]
```diff
@@ -21,10 +21,97 @@ def __init__(self):
...
+                                            )
+            logger.info("---------------------------------------")
+            await asyncio.sleep(60)
+
+    async def _listen_udp(self):
         # Instantiate the UDP listener.
-        udp_listener = AsyncUDPListener(callback=self.handle_payload_summary, port=55673)
+        udp_listener = AsyncUDPListener(
+            callback=self.handle_payload_summary,
+            port=self._settings.udp_broadcast.listen_port,
```
Comment (ch3p4ll3) [inline app/radiosonde_auto_rx_listener.py:113]: GGWP

File: app/settings/__init__.py [hunk 6]
```diff
@@ -33,27 +34,11 @@ def load_settings(cls):
     @classmethod
     def get_default_settings(cls):
         data = {
-            "listener_location": {
```
Comment (ch3p4ll3) [inline app/settings/__init__.py:36]: I don't like this :(

File: app/settings/notification_thresholds.py [hunk 7]
```diff
@@ -4,4 +4,6 @@
 class NotificationThresholds(BaseModel):
     distance_km: PositiveFloat  # Distance threshold in kilometers (e.g., send notification if radiosonde is within 20km)
     altitude_meters: float  # Altitude threshold in meters (e.g., send notification if altitude is below 10,000 meters)
-    landing_point_timeout_minutes: int = 5 # Specifies the duration (in minutes) of inactivity after which the landing point is sent. 0 = Disabled
+    landing_point_timeout_minutes: int = (
```
Comment (ch3p4ll3) [inline app/settings/notification_thresholds.py:7]: why?

File: app/utils.py [hunk 8]
```diff
@@ -26,9 +26,27 @@ async def send_notification(message_body, title):
...
+        """Mapp
+        {'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [1.1515, 43.0685, 8021]}, 'properties': {'id': 'MEA901464', 'type': 'M20', 'startplace': 'Aire Sur Adour\t(FR)', 'frequency': '404.00 MHz', 'report': '2024-12-20 10:34:54z', 'speed': '78 km/h', 'course': '158 °', 'climbing': '-10.0 m/s', 'altitude': '8021 m', 'latitude': '43.0685', 'longitude': '1.1515', 'icon': 'sondeIcon'}}
+        to a RadiosondePayload object
+        """
+        return RadiosondePayload(
+            callsign=json_payload["properties"]["id"],
+            model=json_payload["properties"]["type"],
+            freq=json_payload["properties"]["frequency"],
+            batt=-1,
+            vel_v=json_payload["properties"]["climbing"],
```
Comment (ch3p4ll3) [inline app/utils.py:43]: this should be converted to float

File: app/utils.py [hunk 9]
```diff
@@ -26,9 +26,27 @@ async def send_notification(message_body, title):
...
+            vel_h=json_payload["properties"]["speed"],
```
Comment (ch3p4ll3) [inline app/utils.py:44]: also this

File: app/utils.py [hunk 10]
```diff
@@ -26,9 +26,27 @@ async def send_notification(message_body, title):
...
+            altitude=json_payload["properties"]["altitude"],
```
Comment (ch3p4ll3) [inline app/utils.py:45]: and this

File: app/radiosonde_auto_rx_listener.py [hunk 11]
```diff
@@ -21,10 +21,97 @@ def __init__(self):
...
+                        await asyncio.sleep(60)
+                        continue
+                    data = await response.json()
+                    if "features" in data:
+                        for feature in data["features"]:
+                            if (
+                                "geometry" in feature
+                                and feature["geometry"]["type"] == "Point"
+                            ):
+                                coords = (
```
Comment (ch3p4ll3) [inline app/radiosonde_auto_rx_listener.py:73]: why not using 'latitude': '43.0685', 'longitude': '1.1515', on the root of the json?

File: app/utils.py [hunk 12]
```diff
@@ -26,9 +26,27 @@ async def send_notification(message_body, title):
...
+            latitude=json_payload["geometry"]["coordinates"][1],
```
Comment (ch3p4ll3) [inline app/utils.py:46]: why not `json_payload["latitude"]` ?

File: app/settings/__init__.py (hunk 6 above)
Comment (Thepowa753) [reply app/settings/__init__.py:36]: black formatting.

File: app/settings/notification_thresholds.py (hunk 7 above)
Comment (Thepowa753) [reply app/settings/notification_thresholds.py:7]: black formatting.

File: app/settings/__init__.py (hunk 6 above)
Comment (ch3p4ll3) [reply app/settings/__init__.py:36]: bruh

File: app/settings/notification_thresholds.py (hunk 7 above)
Comment (ch3p4ll3) [reply app/settings/notification_thresholds.py:7]: BRUH

File: app/utils.py (hunk 12 above)
Comment (Thepowa753) [reply app/utils.py:46]: Blah blah blah

File: app/utils.py [hunk 13]
```diff
@@ -26,9 +26,27 @@ async def send_notification(message_body, title):
...
+            longitude=json_payload["geometry"]["coordinates"][0],
```
[the remark this answers is not in the document]
Comment (Thepowa753) [reply app/utils.py:47]: gne gne gne

File: requirements.txt [hunk 14]
```diff
@@ -16,3 +16,5 @@ requests==2.32.3
 requests-oauthlib==2.0.0
 typing_extensions==4.12.2
 urllib3==2.2.3
+aiohttp==3.11.11
+datetime==5.5
```
[the remark this answers is not in the document]
Comment (Thepowa753) [reply requirements.txt:20]: :)

File: app/radiosonde_auto_rx_listener.py (hunk 4 above)
Comment (Thepowa753) [reply app/radiosonde_auto_rx_listener.py:64]: Fixed, ball breaker!

**Comment by ch3p4ll3** [comment]
Well, u don't purge old sondes :(

File: app/radiosonde_auto_rx_listener.py [hunk 15]
```diff
@@ -83,13 +83,40 @@ async def _listen_web(self):
...
+                                                ] = {
+                                                    "notify": True,
+                                                    "landing_notify": False,
+                                                    "altitude": int(
+                                                        feature["properties"][
+                                                            "altitude"
+                                                        ].replace(" m", "")
+                                                    ),
+                                                    "last_update": datetime.now(UTC),
+                                                    "data": Utils.map_json_to_radiosonde_payload(
```
[the remark this answers is not in the document]
Comment (Thepowa753) [reply app/radiosonde_auto_rx_listener.py:105]: To mare omo, too late

File: app/radiosonde_auto_rx_listener.py [hunk 16]
```diff
@@ -21,10 +21,111 @@ def __init__(self):
...
+                                            logger.info(
+                                                f"Radiosonde detected within range and below altitude threshold."
+                                            )
+                                            if (
+                                                self._sondes.get(
+                                                    feature["properties"]["id"]
+                                                )
+                                                is None
+                                            ):
+                                                self._sondes[
```
[the remark this answers is not in the document]
Comment (Thepowa753) [reply app/radiosonde_auto_rx_listener.py:94]: me too, but it's convention! dumbass!

[the remark this answers is not in the document]
Comment (ch3p4ll3) [reply app/radiosonde_auto_rx_listener.py:94]: i hate conventions!

File: app/radiosonde_auto_rx_listener.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ch3p4ll3) [reply app/radiosonde_auto_rx_listener.py:105]: https://www.youtube.com/watch?v=ZSM3w1v-A_Y

File: app/radiosonde_auto_rx_listener.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Thepowa753) [reply app/radiosonde_auto_rx_listener.py:94]: Me too!!! DC!

**Pull request merged**